keanu-project/src/main/java/io/improbable/keanu/plating/loop/Loop.java

```diff
@@ -0,0 +1,137 @@
+ *       cond  |  IF   |
+ *           \ |   |\  |
+ *            AND  | iterate
+ *             |\  |  /
+ *             | \ | /
+ *             |  IF
+ *             |   |
+ *             L   V
+ */
+public class Loop {
+    private Logger log = LoggerFactory.getLogger(this.getClass());
+    public static final VertexLabel VALUE_OUT_LABEL = new VertexLabel("loop_value_out");
+    public static final VertexLabel VALUE_IN_LABEL = PlateBuilder.proxyFor(VALUE_OUT_LABEL);
+    public static final VertexLabel CONDITION_LABEL = new VertexLabel("loop_condition");
+    public static final VertexLabel LOOP_LABEL = new VertexLabel("loop");
+    static final VertexLabel STILL_LOOPING_LABEL = PlateBuilder.proxyFor(LOOP_LABEL);
+    public static final int DEFAULT_MAX_COUNT = 100;
+    private final Plates plates;
+    private final boolean throwWhenMaxCountIsReached;
+
+    /**
+     * package-private because it is intended to be created by the LoopBuilder
+     *
+     * @param plates
+     * @param throwWhenMaxCountIsReached
+     */
+    Loop(Plates plates, boolean throwWhenMaxCountIsReached) {
+        this.plates = plates;
+        this.throwWhenMaxCountIsReached = throwWhenMaxCountIsReached;
+    }
+
+    public Plates getPlates() {
+        return plates;
+    }
+
+    /**
+     * A factory method for creating a loop
+     * It automatically labels the initial state correctly for you.
+     *
+     * @param initialState - a single Vertex that defines the loop's base case.
+     * @param <V>          the input type
+     * @return a builder object
+     */
+    public static <V extends Vertex<?>> LoopBuilder startingFrom(V initialState) {
+        if (initialState.getLabel() == null) {
+            initialState.setLabel(VALUE_OUT_LABEL);
+        }
+        return startingFrom(ImmutableList.of(initialState));
+    }
+
+    /**
+     * A factory method for creating a loop
+     *
+     * @param first  - the first Vertex (mandatory)
+     * @param others - other Vertices (optional)
+     * @param <V>    the input type
+     * @return a builder object
+     */
+    public static <V extends Vertex<?>> LoopBuilder startingFrom(V first, V... others) {
+        return startingFrom(ImmutableList.<V>builder().add(first).add(others).build());
+    }
+
+    /**
+     * A factory method for creating a loop
+     *
+     * @param initialState - the collection of vertices that define the loop's base case
+     * @param <V>          the input type
+     * @return a builder object
+     */
+    public static <V extends Vertex<?>> LoopBuilder startingFrom(Collection<V> initialState) {
+        return new LoopBuilder(initialState);
+    }
+
+    /**
+     * @param <V> the output type
+     * @return the output of the Loop (i.e. the output Vertex from the final Plate)
+     * @throws LoopException if the loop was too short and hit its maximum unrolled size
+     */
+    public <V extends Vertex<? extends Tensor<?>>> V getOutput() throws LoopException {
```

Comment:
Does this have to output a Tensor?

Reply:
Same reasons as for the VertexDictionary in my last PR. Therefore same conclusion: I'll revert it to `Vertex<?>` for now, until such time as we can reproduce the issue (if there is any) with  a ClassCastException.